File: benchmarks/protobench.py
#! /usr/bin/env python

import gc
import sys
import timeit

import google.protobuf
import google_size_pb2
import google_speed_pb2

def get_class_by_name(c):
    parts = c.split('.')
    module = ".".join(parts[:-1])
    m = __import__(module)
    for comp in parts[1:]:
        m = getattr(m, comp)
    return m

class Decoder:
    def __init__(self, msgClass, data):
        self.msgClass = msgClass
        self.data = data

    def __call__(self):
        msg = self.msgClass()
        msg.ParseFromString(self.data)
        return True

class Encoder:
    def __init__(self, msg):
        self.msg = msg

    def __call__(self):
        return self.msg.SerializeToString()

def timeAction(numIterations, action):
    t = timeit.Timer(stmt=action)
    gc.collect()
    return t.timeit(number=numIterations)

def benchmark(descr, action, dataSize):
    minSampleTime = 2
    targetTime = 5

    # exercise it a little, to ensure it is loaded and what not
    timeAction(10, action)

    # find out how many iterations we need to make it run for about 30s
    iterations = 1
    elapsed = timeAction(iterations, action)
    while elapsed < minSampleTime:
        iterations = iterations * 2
        elapsed = timeAction(iterations, action)
    iterations = int(round((targetTime / elapsed) * iterations))

    # now make it run
    elapsed = timeAction(iterations, action)
    print "%s: %d iterations in %.3fs; %.2fMB/s" % \
          (descr, iterations,
           elapsed,
           float(iterations * dataSize) / (elapsed * 1024*1024))
    sys.stdout.flush()

def main(args):
    programName = args[0]
    args = args[1:]
    while len(args) >= 2:
        className = args[0]
        msgFile = args[1]
        args = args[2:]

        print "Benchmarking %s with file %s" % (className, msgFile)
        sys.stdout.flush()

        f = open(msgFile, "rb")
        data = f.read()
        f.close()
        c = get_class_by_name(className)
        d = Decoder(c, data)
        msg = c()
        msg.ParseFromString(data)
        e = Encoder(msg)

        benchmark("Serialize to string", e, len(data))
        benchmark("Deserialize from string", d, len(data))

        print
        sys.stdout.flush()

if __name__ == '__main__':
    main(sys.argv)

File: benchmarks/protobench.cc
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

#include <stdio.h>
#include <string.h>
#include <google_speed.pb.h>
#include <google_size.pb.h>
#include <google/protobuf/message_lite.h>

static void benchmark(const char *, const char *,
                      ::google::protobuf::Message *,
                      unsigned char *, int);
static int read_message(const char *, unsigned char **, int *);

int
main(int argc, char **argv)
{
    benchmarks::SpeedMessage1 *m1;
    benchmarks::SpeedMessage2 *m2;
    benchmarks::SizeMessage1 *m3;
    benchmarks::SizeMessage2 *m4;
    unsigned char *data1;
    int data1Len;
    unsigned char *data2;
    int data2Len;

    if (!read_message("google_message1.dat", &data1, &data1Len))
    {
        fprintf(stderr, "Failed top read \"%s\"\n", "google_message1.dat");
        return 1;
    }
    if (!read_message("google_message2.dat", &data2, &data2Len))
    {
        fprintf(stderr, "Failed top read \"%s\"\n", "google_message2.dat");
        return 1;
    }
    /*
      printf("Read %d bytes into %x\n", data1Len, (unsigned int)data1);
      printf("Read %d bytes into %x\n", data2Len, (unsigned int)data2);
    */


    m1 = new benchmarks::SpeedMessage1();
    m2 = new benchmarks::SpeedMessage2();
    m3 = new benchmarks::SizeMessage1();
    m4 = new benchmarks::SizeMessage2();

    benchmark("SpeedMessage1", "google_message1.dat", m1, data1, data1Len);
    benchmark("SpeedMessage2", "google_message2.dat", m2, data2, data2Len);
    benchmark("SizeMessage1",  "google_message1.dat", m3, data1, data1Len);
    benchmark("SizeMessage2",  "google_message2.dat", m4, data2, data2Len);

    delete m1;
    delete m2;
    delete m3;
    delete m4;
    return 0;
}

#define TIME_ACTION(numIterations, action, result)              \
    {                                                           \
        struct timeval t0, t1, res;                             \
        long i = numIterations;                                 \
        gettimeofday(&t0, NULL);                                \
        for (i = 0; i < numIterations; i++)                     \
        {                                                       \
            action ;                                            \
        }                                                       \
        gettimeofday(&t1, NULL);                                \
        timersub(&t1, &t0, &res);                                       \
        result = (double)(res.tv_sec) + (double)res.tv_usec / 1000000.0; \
    }

#define BENCHMARK(descr, minSampleTime, targetTime, op, dataLen)        \
    {                                                                   \
        double minSampleT = minSampleTime;                              \
        double targetT = targetTime;                                    \
        double elapsed;                                                 \
        long iterations;                                                \
                                                                        \
        iterations = 1;                                                 \
        TIME_ACTION(iterations, op, elapsed);                           \
        while (elapsed < minSampleT)                                    \
        {                                                               \
            iterations *= 2;                                            \
            TIME_ACTION(iterations, op, elapsed);                       \
        }                                                               \
        iterations = (long)((targetT / elapsed) * (double)iterations);  \
                                                                        \
        TIME_ACTION(iterations, op, elapsed);                           \
        printf("%s: %ld iterations in %.3fs; %.2fMB/s\n",               \
               descr, iterations, elapsed,                              \
               ((double)iterations*(double)dataLen) /                   \
               (elapsed * 1024.0 * 1024.0));                            \
    }

static void
benchmark(const char *classname, const char *filename,
          ::google::protobuf::Message *m,
          unsigned char *src, int len)
{
    unsigned char dest[len+1000];
    int destLen = len+100;
    double minSampleTime = 2.0;
    double targetTime = 30.0;

    if (!m->ParseFromArray(src, len))
    {
        printf("parsing message...failed :(\n");
        return;
    }
    if (!m->SerializeToArray(dest, destLen))
    {
        printf("serializing message...failed :(\n");
        return;
    }

    printf("Benchmarking %s with file %s\n", classname, filename);
    BENCHMARK("Serialize to array", minSampleTime, targetTime,
              (m->SerializeToArray(dest,destLen)),
              len);

    BENCHMARK("Deserialize from array", minSampleTime, targetTime,
              (m->ParseFromArray(src,len)),
              len);
    printf("\n");
}

static int
read_message(const char *filename, unsigned char **dest, int *len)
{
    struct stat st;
    int fd, numRemaining, numRead, error;

    if (stat(filename, &st)==-1)
    {
        return 0;
    }

    *len = st.st_size;
    if ((*dest = (unsigned char *)malloc(st.st_size)) == NULL)
        return 0;

    if ((fd = open(filename, O_RDONLY)) == -1)
    {
        free(*dest);
        return 0;
    }

    numRemaining = (int)st.st_size;
    numRead = 0;
    error = 0;
    do
    {
        int res;
        switch (res = read(fd, (*dest) + numRead, numRemaining))
        {
            case 0:
                numRemaining = 0;
                break;
            case -1:
                switch (errno) {
                    case EINTR: break;
                    default: error = 1;
                }
                break;
            default:
                numRemaining -= (int)res;
                numRead += (int)res;
                break;
        }
    } while (numRemaining > 0 && !error);

    close(fd);

    if (error)
    {
        free(*dest);
        return 0;
    }

    return 1;
}
